### DISCUSSION TITLE
Audio Device/interface support:  for AIS decoding

### DISCUSSION THREAD

**Discussion by rkarikari** [jvde-github/AIS-catcher#124] [General]
Hello ,
Thanks for the great work done with "AIS catcher".  Really enjoying it's performance (on my RTL_SDR).

Is there a way to get AIS-catcher to decode from an audio device  ?
( IF from USB port of Amateur Radio  ) ?

My "ham radio"  has a USB output that appears on my Mac as an "audio  interface".
This works with  other SDR software ( SDRangel via  "audio Input" ) to decode AIS packets from my radio.

In other words.......
I decode AIS packets from my Amatuer Radio (Kenwood TH-D74) via SDRangel's "audio Input" .
I would prefer to use AIS-catcher ..... but cannot figure out how to get this done.

Is it possible to select an audio interface instead of a file ?

Thanks so much.

regards
Rich

**Comment by jvde-github** [comment]
hello rich,

There is no direct audio driver but maybe there are other options. What is the audio format from the device, is it FM decoded or raw IQ data? What is the frequency and bandwidth?

Sorry, I am not familiar with the device mentioned. Ideal would be to share a very short recording with a few messages for testing.

Thanks, Jasper

**Reply by rkarikari** [reply]
Hello Jasper,

Thanks for the quick response. 

The device is simply a radio with raw output capability.
The radio can do both raw IQ and  modulated output ( FM, USB, CW, AM etc)
I set it to raw IQ on SDRangel and it decodes AIS just fine.

The bandwidth is 16k so I can decode from 1 channel at a time... either 161.975 or 162.025.... not both.

THanks again.

**Reply by jvde-github** [reply]
Perfect. The approach I am thinking about is to use a program like sox to read from an audio device and then pipe that audio into AIS-catcher for decoding. We need to add a new decoding model though that reads only from one channel at a sample rate of 16k. I can add that if useful.

**Reply by rkarikari** [reply]
Yes, that would be excellent!!!

Very much appreciated.

**Comment by jvde-github** [comment]
I created a test version in the 'single' branche.

Clone via:
```
git clone https://github.com/jvde-github/AIS-catcher.git -b single
```
Then build this software. You can use this to decode IQ input at 16k in float 32 that is coming in on stdin:
```
cat iq16k.raw | AIS-catcher -c X A -r cf32 . -s 16k -v
```
In this example I am decoding the input in a file iq16k.raw and show the NMEA lines assuming the input is one channel with designation A (`-c X A`).

My suggestion is to somehow dump the audio to stdout instead of reading from file... The file is btw [here](https://github.com/jvde-github/AIS-catcher/blob/088e1ab96edbf22beb17bb72fc34ef7fbf4acc96/iq16k.raw)

And if the file is in 16 bit signed int:
```
cat iq16k_cs16.raw | AIS-catcher -r cs16 . -c X A -s 16k
```
File can be found [here](https://github.com/jvde-github/AIS-catcher/blob/c42da88e845ed9a155af0d59e057c2e58bb33517/iq16k_cs16.raw)

**Comment by rkarikari** [comment]
I followed your instructions.. and compiled the "single" branch ...
then tested it with your sample audio files and it decodes them (both)  just fine.

However, I could not get it to decode from my radio.
  I see the audio is being routed to it fine... but no decodes occur.

  - I tried directly from the audio interface ...
     `sox -t coreaudio "ADC stream IN" -t raw -b 16 -e unsigned -r 96k -  | AIS-catcher -s 16K -r cf32 . -c X A -v`
     `sox -t coreaudio "ADC stream IN" -t raw -b 16 -e signed -r 96k -  | AIS-catcher -s 16K -r cs16 . -c X A -v`
             .... along with various other changes.  Could not get any to work.

  - I saved raw IQ files from SDRangel and tried them as well
             `cat test.raw | AIS-catcher -r cs16 . -s 16k -c X A -v`

None of them worked.

My radio's audio output is :  
   -  48k 16-bit mono  

In SDRangel,
  -  The IQ sample rate is: 48k. ( Auto detected.) from my radio.
        -     it refers to it as : Baseband IQ sample rate.
        -      but the AIS decoder bandwidth is set at 16k.
                     This is what works and decodes fine.

Not sure what I'm missing... still scratching my head.

**Comment by jvde-github** [comment]
Hello,

I think the SDRangel file should then be decoded with -s 48k as this is the file format.

On your core audio example, you have:
```
sox -t coreaudio "ADC stream IN" -t raw -b 16 -e signed -r 96k -  | AIS-catcher -s 16K -r cs16 . -c X A -v
```
Which creates output at 96k in signed integer format. I suggest you put the output in 16k (-r 16k), then the sox output looks more compatible with the AIS-catcher expected input (-s 16k). AIS-catcher needs to know the sample rate of the input, not the bandwidth. One other point is that AIS-catcher with -c X can only handle now samples rates <= 48k, so that is why I suggest to change the -r. On another point, the audio is stereo and represents IQ samples?

On your radio output, which is 48k mono. Are you sure this is raw IQ output? The problem is that for IQ format I expected two channels (for I and Q). Perhaps this is FM demodulated already?

**Comment by jvde-github** [comment]
Feel free to share any recordings here or by my email if you want me to have a closer look.

**Comment by rkarikari** [comment]
Thanks so much for the feedback and suggestions.

I am 99% certain it is not FM modulated/demodulated.
It is described as "IF" and is output at a stage before  any demodulation.

This is what the interface and spectrum/waterfall look like in SDRangel. 
The highlighted portion is the 16k width for decoding.

<img width="381" alt="Screenshot 2023-06-30 at 12 05 15 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/c8baca19-0ac6-4f64-8859-151f3510d2d1">
<img width="392" alt="Screenshot 2023-06-30 at 1 14 36 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/313d5cd1-63ae-4eb4-a6bd-d0aedd1cd1e7">



I'll try to get a small file recording to post soon. 


Also, I had -r at 96k because I was getting "FILE timeout" errors at any other lower setting. 

<img width="977" alt="Screenshot 2023-06-30 at 12 56 58 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/586ef4dc-5102-4ebc-963a-f096b37c8e24">

**Comment by rkarikari** [comment]
Please find below the raw file ..... and how it was recorded.
Hopefully, it provides some info.

[test.raw.zip](https://github.com/jvde-github/AIS-catcher/files/11914992/test.raw.zip)
<img width="873" alt="Screenshot 2023-06-30 at 9 25 28 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/80030bc1-c650-4659-be1c-9a6ed1da49bd">

Please find above the raw file .

**Reply by jvde-github** [reply]
It is slowly getting clearer. It is IQ input but only for the I part so some additional processing is required. It is interesting enough so will have a look. It seems to be in signed 16 bit integers at 16k.

I think this sampling rate is too low btw as the max bandwidth is 8k which is not wide enough for AIS. So if it is mono I suggest to keep it at 48k sample rate for now (-r 48000).

Thank you

**Comment by rkarikari** [comment]
[AIS_2023-07-03T22_31_47_488.wav.zip](https://github.com/jvde-github/AIS-catcher/files/11943202/AIS_2023-07-03T22_31_47_488.wav.zip)

Playing with an IQ file (see attached above) saved from SDRangel, I got the following lines below decoded.
This is using the "main" branch

<img width="670" alt="Screenshot 2023-07-04 at 12 25 50 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/718b8dee-7013-4919-ba7c-89505b7a29e3">


but SDRangel got another set.... from the same file.... different MMSI numbers.

<img width="935" alt="Screenshot 2023-07-04 at 12 28 38 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/e8dd0623-3a97-4941-85e5-91d391196dbd">

All the vessels (from both decodes) were actually in range.... very interesting.

**Comment by jvde-github** [comment]
I think the AIS-catcher results were a coincidence as the file-format is not cs8 at 96k. I had a look at the wave-file header using `mediainfo` and it suggests that it is in signed int16 IQ format at 48k and only for one AIS channel (note that SDRangel does not mention channel either). Unfortunately, AIS-catcher seems to struggle with this WAV-file format, so I convert it first with `soxr` and then decode with one-channel version of AIS-catcher:  
```
sox ~/Downloads/AIS_2023-07-03T22_31_47_488.wav -t raw - | AIS-catcher  -r cs16 . -c X -s 48k -v
AIS-catcher (build Jun 30 2023) v0.48-38-g975ed64
(C) Copyright 2021-2023 jvde-github and other contributors
This is free software; see the source for copying conditions.There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Device    : FILE-RAW
Settings  : rate 48K format CS16 file .
Model #0 -> 1: [AIS engine v0.48] ps_ema ON afc_wide ON droop ON fp_ds OFF 
!AIVDM,1,1,,X,33MAl8Q000008Pj3:vVL;VoR0Dp:,0*0F ( MSG: 3, REPEAT: 0, MMSI: 232027170)
!AIVDM,1,1,,X,19NSRH0vAsOvcjN2mLB9i7f800S8,0*39 ( MSG: 1, REPEAT: 0, MMSI: 636019296)
!AIVDM,1,1,,X,18KSrv802HP1GGR3:Ru:a`Q<0853,0*1D ( MSG: 1, REPEAT: 0, MMSI: 565771000)
!AIVDM,1,1,,X,38J=k@1000P0@UF3:SH<iVhJ0000,0*69 ( MSG: 3, REPEAT: 0, MMSI: 564360000)
!AIVDM,1,1,,X,18KsRd002LwsIdR2gLDICWNp0@Bn,0*5F ( MSG: 1, REPEAT: 0, MMSI: 566158000)
!AIVDM,1,1,,X,18KsRd002LwsINT2gK?9BoO@06j0,0*77 ( MSG: 1, REPEAT: 0, MMSI: 566158000)
!AIVDM,1,1,,X,19NSRH02ArOvc;t2mJIIfGiF06j0,0*53 ( MSG: 1, REPEAT: 0, MMSI: 636019296)
!AIVDM,2,1,0,X,59NSd2`2GNUwDhI3>20m<>045T62222222222216=l0NE5jV0Vm1k1Up,0*77 ( MSG: 5, REPEAT: 0, MMSI: 636021770)
!AIVDM,2,2,0,X,888888888888880,2*3D ( MSG: 5, REPEAT: 0, MMSI: 636021770)
!AIVDM,1,1,,X,198VjT@01>wvEB>2nD<aVWwf26j0,0*2B ( MSG: 1, REPEAT: 0, MMSI: 613003921)
!AIVDM,1,1,,X,18KsRd002LwsI1N2gHuqBWN:06j4,0*54 ( MSG: 1, REPEAT: 0, MMSI: 566158000)
!AIVDM,1,1,,X,19NSRH0vArOvbRV2mHLIg7fh0@?=,0*24 ( MSG: 1, REPEAT: 0, MMSI: 636019296)
!AIVDM,2,1,1,X,58KsRd800003UKWGC60l4E9<f0=DpDpF22222216GprJ;5mnNH1j51CO,0*2B ( MSG: 5, REPEAT: 0, MMSI: 566158000)
!AIVDM,2,2,1,X,ggS@D53L<H88880,2*55 ( MSG: 5, REPEAT: 0, MMSI: 566158000)
!AIVDM,1,1,,X,198VjT@01=wvDq>2nC69qow226j4,0*00 ( MSG: 1, REPEAT: 0, MMSI: 613003921)
!AIVDM,1,1,,X,18KSrv802IP1EGf3:Sj:dHPL0HI`,0*37 ( MSG: 1, REPEAT: 0, MMSI: 565771000)
!AIVDM,1,1,,X,18KsRd002LwsHHV2gEh9BGOB00Ru,0*7A ( MSG: 1, REPEAT: 0, MMSI: 566158000)
!AIVDM,1,1,,X,19NSRH0v1rOvb@22mGPqdWgF00Sq,0*15 ( MSG: 1, REPEAT: 0, MMSI: 636019296)
!AIVDM,2,1,2,X,58J=k@02AF3KUKW;KV0l4E9<f0<4h485:2222218GhsK:5j>0KPR`0kk,0*0E ( MSG: 5, REPEAT: 0, MMSI: 564360000)
!AIVDM,2,2,2,X,wQj851CH8888880,2*44 ( MSG: 5, REPEAT: 0, MMSI: 564360000)
!AIVDM,1,1,,X,18KsRd002LwsH:V2gDbIBoO`06j4,0*06 ( MSG: 1, REPEAT: 0, MMSI: 566158000)
----------------------
[AIS engine v0.48 #0-0]                 total: 18 msgs
```

**Comment by jvde-github** [comment]
I pushed a small change to the 'single' branche so this now works directly:
```
AIS-catcher -w AIS_2023-07-03T22_31_47_488.wav -c X -v
```

**Comment by rkarikari** [comment]
Thank you so much for your time and patience. Greatly appreciated.
I'm able to decode the file now....  so much more decoded compared to SDRangel.   :)

I tried directly decoding from the audio input with no success (no decodes).
I recorded from the audio source using sox with settings to mimic the file format successfully decoded.

[test.wav.zip](https://github.com/jvde-github/AIS-catcher/files/11950284/test.wav.zip)


As seen below, the are both with the same details. 
<img width="515" alt="wavs compared" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/65413751-7735-486b-9d01-16c2ff3e66c6">

However, test.wav recorded with sox does not decode as AIS..wav does.

I compared them in SDRangel and noticed the following...

Left: SDRangel (AIS..wav )      ---------------------------------------------------------          Right: Sox (test.wav )

<img width="389" alt="SDRangel wav" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/46a576e4-dfe1-4586-b5bd-93441065a0dd">
<img width="388" alt="sox wav shift" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/579dfaa5-d8c1-4c99-82e1-1f3156f5a657">

Left: SDRangel wav recorded    Right: Sox wav recorded.
 
 To decode test.wav in SDRangel, I have to shift the highlight from the middle to the left.


Is it possible to do this in AIS-catcher ?

Any suggestions ?

Again, Thank you so much.

**Comment by jvde-github** [comment]
Great! If you can record you can directly pipe the data into AIS-catcher and can run it as a proper receiver with web interface. So worthwhile to see if we can make it work.

There are some brilliant command line DSP utilities that allow you to shift the frequency of a signal: https://github.com/ha7ilm/csdr

So a command like this works on your file:

```
sox test.wav -t raw -e floating-point -b 32  -c 2 - | csdr shift_math_cc +0.25 | AIS-catcher -r cf32 . -c X -s 48k -v -M D
AIS-catcher (build Jul  4 2023) v0.48-38-g975ed64
(C) Copyright 2021-2023 jvde-github and other contributors
This is free software; see the source for copying conditions.There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Device    : FILE-RAW
Settings  : rate 48K format CF32 file .
Model #0 -> 1: [AIS engine v0.48] ps_ema ON afc_wide ON droop ON fp_ds OFF 
!AIVDM,1,1,,X,15UJPh000700Hc43;BS1CGGP0L1:,0*48 ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -41.193, ppm: -0.868056)
!AIVDM,1,1,,X,15UJPh000800HcH3;Be13oH@0<1:,0*3A ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -40.5298, ppm: -0.578704)
!AIVDM,1,1,,X,35?VQ0100000AiP3;>a5FVLB0000,0*34 ( MSG: 3, REPEAT: 0, MMSI: 351904000, signalpower: -46.9421, ppm: 0.289352)
!AIVDM,1,1,,X,15UJPh000800Hd03;Bo18oJt0D1:,0*0A ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -41.9835, ppm: 0.578704)
!AIVDM,1,1,,X,19NS>`@019P0Ogd3<OmdBIe>00Sh,0*02 ( MSG: 1, REPEAT: 0, MMSI: 636014241, signalpower: -40.1999, ppm: 0)
!AIVDM,1,1,,X,15UJPh000700HdD3;BvQ<WKP0HL>,0*64 ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -41.9462, ppm: 0.578704)
!AIVDM,1,1,,X,19NS>`@018P0OV03<Q5t?Id005aH,0*55 ( MSG: 1, REPEAT: 0, MMSI: 636014241, signalpower: -40.0622, ppm: 0)
!AIVDM,1,1,,X,15UJPh000700He@3;C8PuWJ@084T,0*15 ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -42.8015, ppm: 0.578704)
!AIVDM,1,1,,X,14cw>0000K003:V3>7eJc`L`00S9,0*54 ( MSG: 1, REPEAT: 0, MMSI: 314560000, signalpower: -44.3426, ppm: 1.44676)
!AIVDM,1,1,,X,15UJPh000700HeT3;CBPnWJt0@AJ,0*47 ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -42.9707, ppm: -0.868056)
!AIVDM,1,1,,X,377fa`1003P05Q:3;quqvnPR0Dg:,0*19 ( MSG: 3, REPEAT: 0, MMSI: 477866400, signalpower: -42.8265, ppm: -0.868056)
!AIVDM,1,1,,X,19NS>`@018P0OCf3<SD<>qc@0D0l,0*67 ( MSG: 1, REPEAT: 0, MMSI: 636014241, signalpower: -44.1575, ppm: 0)
!AIVDM,1,1,,X,15UJPh000600Hf<3;CLPfoKP0D1:,0*32 ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -42.709, ppm: -0.578704)
!AIVDM,1,1,,X,14cw>0000G003283>7h:f8On0<1;,0*7F ( MSG: 1, REPEAT: 0, MMSI: 314560000, signalpower: -48.3725, ppm: -1.73611)
!AIVDM,1,1,,X,35UJPh100500Hf<3;CT0PWH@0Bsb,0*4B ( MSG: 3, REPEAT: 0, MMSI: 374776000, signalpower: -41.8942, ppm: 0.578704)
!AIVDM,1,1,,X,15UJPh100500HfP3;CmP9GEP0000,0*7B ( MSG: 1, REPEAT: 0, MMSI: 374776000, signalpower: -42.979, ppm: 0.578704)
!AIVDM,1,1,,X,19NS>`@018P0NfJ3<WV<>qen0@20,0*0B ( MSG: 1, REPEAT: 0, MMSI: 636014241, signalpower: -46.8737, ppm: -0.578704)
```
The signal is at 48K and the shift of 0.25 shifts it by 12K which is roughly what you do in your screenshots.

Now we need to think how to easily do this shifting in AIS-catcher without CSDR. Will be fairly easy to implement so will add in the coming days. 

Then you would be able to do something like record audio from the device and write to the stdout and at the same time read it into AIS-catcher.

**Comment by rkarikari** [comment]
Unfortunately,  cannot find a macos version of csdr.

No success  compiling  it either... 

Is there an alternative for macos ?

**Comment by jvde-github** [comment]
you can put below in a file rotate.cpp and then compile with:
```
clang++ -std=c++11 -o rotate rotate.cpp -O3
```
the command becomes:
```
sox test.wav -t raw -e floating-point -b 32  -c 2 - | ./rotate | AIS-catcher -r cf32 . -c X -s 48k -v 
```

Bit slow but does the trick until it is included in AIS-catcher. Let me know if it works for you

```console
#include <iostream>
#include <cstdint>
#include <complex>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
   
    // Use std::cin in binary mode
    std::cin.unsetf(std::ios::skipws);

    // The repeating sequence of multipliers 1, 1i, -1, -1i
    std::vector<std::complex<float>> multipliers = {
        std::complex<float>(1, 0),
        std::complex<float>(0, 1),
        std::complex<float>(-1, 0),
        std::complex<float>(0, -1)
    };
    int index = 0;

    // Loop until end of input
    while (true) {
        float re, im;
       
        // Read two 32-bit values (real and imaginary parts) from stdin
        std::cin.read(reinterpret_cast<char*>(&re), sizeof(re));
        std::cin.read(reinterpret_cast<char*>(&im), sizeof(im));
       
        // Check if reading was successful
        if (!std::cin) {
            break; // break out of loop if failed to read
        }
       
        // Create a complex number and multiply it by the next multiplier
        std::complex<float> c(re, im);
        std::complex<float> result = c * multipliers[index];
       
        // Store the real and imaginary parts in separate variables
        float result_real = result.real();
        float result_imag = result.imag();
       
        // Write the resulting complex number in binary format to stdout
        std::cout.write(reinterpret_cast<const char*>(&result_real), sizeof(result_real));
        std::cout.write(reinterpret_cast<const char*>(&result_imag), sizeof(result_imag));
       
        // Move to the next multiplier in the sequence
        index = (index + 1) % 4;
    }

    return 0;
}
```

**Comment by rkarikari** [comment]
It works well with the sox recorded files..... 
but when replaced with the sound interface, it's hit and miss (mostly miss)  

.... as if it's off-frequency.

<img width="1229" alt="Screenshot 2023-07-04 at 8 38 12 PM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/158316af-0a39-4c97-ae6d-961c22428195">

So I set them up to run simultaneously and it's getting a very very low fraction of what SDRangel decodes.....

**Comment by jvde-github** [comment]
what is he command you are using for creating the sox recorded files compared to getting it from the sound interface?

**Comment by rkarikari** [comment]
The last uploaded  file  was created thus.....

           ` sox -t coreaudio "ADC stream IN" -t wav -b 16 -e signed  -c 2 test.wav  gain +10 trim 0 40  ` 

..... for a 40 sec recording

**Comment by jvde-github** [comment]
in your live example, where you stream into AIS-catcher, would it help that after the "sox" you add "-c 2 -b 16" to force stereo capturing. Somehow the capturing is in mono in the screenshot and it does not work. The -c later converts it then to stereo but the capturing from the sound card needs to be stereo.

**Comment by rkarikari** [comment]
<img width="1185" alt="Screenshot 2023-07-05 at 7 37 04 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/82141219-a896-4f0e-bcea-99246ece57b5">

as shown above, it says cant set 2 channels... using 1.

**Reply by jvde-github** [reply]
Ok, and what is the output of?
```
sox -t coreaudio "ADC stream IN" -t wav -b 16 -e signed  -c 2 test.wav  gain +10 trim 0 40  
```
Weird that it works when writing to a file first....

**Comment by rkarikari** [comment]
<img width="948" alt="Screenshot 2023-07-05 at 8 47 46 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/be7bf18c-e9c7-489c-817c-8520507ff7be">
<img width="596" alt="Screenshot 2023-07-05 at 8 51 39 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/0ca80610-4ec3-4330-89a2-b88c84370afe">


[test.wav.zip](https://github.com/jvde-github/AIS-catcher/files/11955923/test.wav.zip)

Nothing decoded from ais-catcher..



but SDRangel gets this..


<img width="1534" alt="Screenshot 2023-07-05 at 8 55 48 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/29ee66ef-80da-4493-846c-99b4fc7b3e9e">

**Comment by rkarikari** [comment]
I managed to get csdr compiled.... and playing with the shift...

<img width="1201" alt="Screenshot 2023-07-05 at 11 05 05 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/10256f4c-4149-47b0-9fdd-8fc0a00567d7">

+0.05

**Comment by rkarikari** [comment]
Thank you so very much Jasper. 

I've been experimenting  and  found the best rate at -0.25 and ( +0.25 as you suggested ). 

I ran multiple sessions concurrently ....  ( one as control -0.25 ) and the other sessions with varying  "shift" magnitude and compared the decode rate/performance. 

.. and csdr is truly a great little tool... I've enjoyed using it. (  ...though I had to butcher it to get it to compile on macOS ).

I lost my internet connection for a while nd noticed AIS-catcher does not work as normal ( wont display web pages ) offline.


Is there a way to " cache " it locally so it works as normal when offline?
Is it possible to extend the statistics retention to a year ? 

Again..  Thank you !!!!

**Comment by jvde-github** [comment]
Hi, thanks, The website does work offline in itself but it uses several packages online (graph.js, leaflet, etc) for plotting and mapping. I like that the program is just one file. That keeps it easy for people to install so need to think a little bit how to resolve this. One approach is to collapse all the dependencies in the program but that makes it larger and still does not get you the maps offline.

Have you ever looked at OpenCPN that runs locally?

**Comment by rkarikari** [comment]
I have looked at OpenCPN  but it does not suit my needs.
 No performance data as far as I can tell. .... and my primary use of AIS catcher is for "benchmarking" / performance comparisons. 

I just wish I could somehow keep a local copy of all the "online referenced data" .
A "local cache" of some sort .    

I'm just not a fan on being held hostage by commercial services  .   

:)

**Reply by jvde-github** [reply]
Have you tried the BACKUP option:

https://github.com/jvde-github/AIS-catcher#plots

This will save the performance statistics in a file and will retrieve that when the program is restarted. So, it will become persistent.

**Reply by rkarikari** [reply]
Yes, I use it.... makes a world of difference for review.

Was wondering how long the data is maintained.

<img width="688" alt="Screenshot 2023-07-09 at 7 16 54 AM" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/a525a147-1062-40e9-8d49-8046efad6ad0">

At 16 days now (after multiple restarts ) but was wondering if it will persist for at least a year ?

:)

**Reply by jvde-github** [reply]
90 days.... unfortunately it is not easy to change
